pageserver/src/http/routes.rs

```diff
@@ -0,0 +1,164 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use hyper::header;
+use hyper::StatusCode;
+use hyper::{Body, Request, Response, Uri};
+use routerify::Middleware;
+use routerify::{ext::RequestExt, RouterBuilder};
+use zenith_utils::auth::JwtAuth;
+use zenith_utils::http::endpoint::attach_openapi_ui;
+use zenith_utils::http::endpoint::auth_middleware;
+use zenith_utils::http::endpoint::check_permission;
+use zenith_utils::http::endpoint::AuthProvider;
+use zenith_utils::http::error::ApiError;
+use zenith_utils::http::{
+    endpoint,
+    error::HttpErrorBody,
+    json::{json_request, json_response},
+};
+
+use super::models::BranchCreateRequest;
+use super::models::TenantCreateRequest;
+use crate::page_cache;
+use crate::{
+    branches::{self},
+    PageServerConf, ZTenantId,
+};
+
+#[derive(Debug)]
+struct State {
+    conf: &'static PageServerConf,
+    auth: Arc<Option<JwtAuth>>,
+    whitelist_routes: Vec<Uri>,
+}
+
+impl State {
+    fn new(conf: &'static PageServerConf, auth: Arc<Option<JwtAuth>>) -> Self {
+        let whitelist_routes = ["/v1/status", "/v1/doc", "/swagger.yml"]
+            .iter()
+            .map(|v| v.parse().unwrap())
+            .collect::<Vec<_>>();
+        Self {
+            conf,
+            auth,
+            whitelist_routes,
+        }
+    }
+}
+
+impl AuthProvider for State {
+    fn provide_auth(&self, req: &Request<Body>) -> Arc<Option<JwtAuth>> {
+        if self.whitelist_routes.contains(req.uri()) {
+            Arc::new(None)
+        } else {
+            self.auth.clone()
+        }
+    }
+}
+
+// healthcheck handler
+async fn status_handler(_: Request<Body>) -> Result<Response<Body>, ApiError> {
+    Ok(Response::builder()
+        .status(StatusCode::OK)
+        .header(header::CONTENT_TYPE, "application/json")
+        .body(Body::from("{}"))
+        .map_err(ApiError::from_err)?)
+}
+
+async fn branch_create_handler(mut request: Request<Body>) -> Result<Response<Body>, ApiError> {
+    let state = request.data::<Arc<State>>().unwrap().clone();
```

Comment:
It's possible to rewrite this function to get rid of this `.clone()`. Apparently, in these async methods you only ever reference `PageServerConf`, so I'd introduce a small helper function:

```rust
fn get_config(request: &Request<Body>) -> &'static PageServerConf {
    request.data::<Arc<State>>().unwrap().conf
}
```

This would encapsulate the fragile `.data::<...>()` logic and make config extraction less bulky at the same time, e.g.

```rust
async fn tenant_list_handler(request: Request<Body>) -> Result<Response<Body>, ApiError> {
    // check for management permission
    check_permission(&request, None)?;

    let response_data =
        tokio::task::spawn_blocking(move || crate::branches::get_tenants(get_config(&request)))
            .await
            .map_err(ApiError::from_err)??;
    Ok(json_response(StatusCode::OK, response_data)?)
}
```

Reply:
I've tried this approach but decided against it because in case new field used in handlers will be added to State, or config reference becomes non static this would require copying anyway, so state extraction is more general. But I agree with separation of this logic into separate function, this looks much clearer

Reply:
> because in case new field used in handlers will be added to State, or config reference becomes non static this would require copying anyway

I think you're overthinking. First of all, I see no problem with new fields: either you can just reference them and hence no changes are needed, or you do need to change something if an owned value is needed, but then again, it's inevitable: nothing is forever. You can't foresee everything, hence IMO it's better to stick with simple yet robust designs and replace them when they no longer fit.

Speaking of a "static" config reference: it's only needed for a function which accepts `'static PageServerConfig`, otherwise I would've chosen `& 'a PageServerConfig`.

Reply:
> you can just reference them 

If a understand it correctly this will already require changes, or addition of new function e.g. get_new_field(req)

Concerns related to static references are about passing them to other threads, I suppose launching tokio::spawn_blocking, and do not awaiting it immediately to let it run in the background can require either an owned object (copy) or static reference. But maybe this is also overthinking :)

Reply:
> If a understand it correctly this will already require changes, or addition of new function e.g. get_new_field(req)

Why would it? I created a new function on a whim, because it occurred to me that it looked prettier. But you could just use `get_state`.

Reply:
> Concerns related to static references are about passing them to other threads, I suppose launching tokio::spawn_blocking

Think of it this way: does one piece of code requiring `Arc<State>` (for whatever reason) mean that every other method must suffer the `.clone()` penalty?

Reply:
> Why would it?

You said no changes required, and with get_state indeed no changes required, so I'm just groping

> Think of it this way

I think now this is just about ergonomics and penalty is not that large, but anyway I think this is indeed premature, so when there will be a requirement (maybe never) to have an owned value, then this code should evolve. So for now I think it's a good option to stick with `get_state`, what do you think?

Reply:
> it's a good option to stick with `get_state`

Yeah, as long as it returns a reference to `State`, why not?